File: src/net/dinkla/raytracer/materials/SVPhong.java
package net.dinkla.raytracer.materials;

import net.dinkla.raytracer.brdf.SVGlossySpecular;
import net.dinkla.raytracer.colors.ColorAccumulator;
import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.lights.AreaLight;
import net.dinkla.raytracer.lights.Light;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.textures.Texture;
import net.dinkla.raytracer.worlds.World;

import java.util.List;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 03.06.2010
 * Time: 18:47:14
 * To change this template use File | Settings | File Templates.
 */
public class SVPhong<C extends Color> extends SVMatte<C> {

    public SVGlossySpecular<C> specularBrdf;

    public SVPhong() {
        super();
        specularBrdf = new SVGlossySpecular<C>();
    }

    public void setKs(double ks) {
        specularBrdf.ks = ks;
    }

    public void setExp(double exp) {
        specularBrdf.exp = exp;
    }

    public void setCs(final Texture<C> cs) {
        specularBrdf.cs = cs;
    }

    @Override
    public C shade(World<C> world, Shade sr) {
        Vector3D wo = sr.ray.getD().negate();
        C L = getAmbientColor(world, sr, wo);
        for (Light light : world.getLights()) {
            Vector3D wi = light.getDirection(sr);
            double nDotWi = sr.getNormal().dot(wi);
            if (nDotWi > 0) {
                boolean inShadow = false;
                if (light.shadows) {
                    Ray shadowRay = new Ray(sr.getHitPoint(), wi);
                    inShadow = light.inShadow(world, shadowRay, sr);
                }
                if (!inShadow) {
                    Color fd = diffuseBrdf.f(sr, wo, wi);
                    Color fs = specularBrdf.f(sr, wo, wi);
                    Color l = light.L(world, sr);
                    Color fdfslndotwi = fd.plus(fs).mult(l).mult(nDotWi);
                    L = (C) L.plus(fdfslndotwi);
                }
            }
        }
        return L;
    }


    @Override
    public C areaLightShade(World<C> world, Shade sr) {
        Vector3D wo = sr.ray.getD().negate();
        C L = getAmbientColor(world, sr, wo);
        ColorAccumulator S = new ColorAccumulator();
        for (Light light1 : world.getLights()) {
            if (light1 instanceof AreaLight) {
                AreaLight light = (AreaLight) light1;
                List<AreaLight.Sample> ls = light.getSamples(sr);
                for (AreaLight.Sample sample : ls) {
                    double nDotWi = sample.wi.dot(sr.getNormal());
                    if (nDotWi > 0) {
                        boolean inShadow = false;
                        if (light.shadows) {
                            Ray shadowRay = new Ray(sr.getHitPoint(), sample.wi);
                            inShadow = light.inShadow(world, shadowRay, sr, sample);
                        }
                        if (!inShadow) {
                            Color fd = diffuseBrdf.f(sr, wo, sample.wi);
                            Color fs = specularBrdf.f(sr, wo, sample.wi);
                            Color l = light.L(world, sr, sample);
                            Color fsfslndotwi = fd.plus(fs).mult(l).mult(nDotWi);
                            // TODO: hier ist der Unterschied zu shade()
                            double f1 = light.G(sr, sample) / light.pdf(sr);
                            Color T = fsfslndotwi.mult(f1);
                            S.plus(T);
                        }
                    }
                }
            }
        }
        L = (C) L.plus(S.getAverage());
        return L;
    }

    @Override
    public C getLe(Shade sr) {
        // TODO
        return (C) specularBrdf.cs.getColor(sr).mult(specularBrdf.ks);
    }
}

File: src/main/kotlin/net/dinkla/raytracer/materials/SVPhong.kt
package net.dinkla.raytracer.materials

import net.dinkla.raytracer.brdf.SVGlossySpecular
import net.dinkla.raytracer.colors.ColorAccumulator
import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.lights.AreaLight
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.textures.Texture
import net.dinkla.raytracer.worlds.World

class SVPhong : SVMatte() {

    var specularBrdf: SVGlossySpecular

    init {
        specularBrdf = SVGlossySpecular()
    }

    fun setKs(ks: Double) {
        specularBrdf.ks = ks
    }

    fun setExp(exp: Double) {
        specularBrdf.exp = exp
    }

    fun setCs(cs: Texture) {
        specularBrdf.cs = cs
    }

    override fun shade(world: World, sr: Shade): Color {
        val wo = sr.ray.d.negate()
        var L = getAmbientColor(world, sr, wo)
        for (light in world.lights) {
            val wi = light.getDirection(sr)
            val nDotWi = sr.normal.dot(wi)
            if (nDotWi > 0) {
                var inShadow = false
                if (light.shadows) {
                    val shadowRay = Ray(sr.hitPoint, wi)
                    inShadow = light.inShadow(world, shadowRay, sr)
                }
                if (!inShadow) {
                    val fd = diffuseBrdf.f(sr, wo, wi)
                    val fs = specularBrdf.f(sr, wo, wi)
                    val l = light.L(world, sr)
                    val fdfslndotwi = fd.plus(fs).times(l).times(nDotWi)
                    L = L.plus(fdfslndotwi)
                }
            }
        }
        return L
    }


    override fun areaLightShade(world: World, sr: Shade): Color {
        val wo = sr.ray.d.negate()
        var L = getAmbientColor(world, sr, wo)
        val S = ColorAccumulator()
        for (light1 in world.lights) {
            if (light1 is AreaLight) {
                val ls = light1.getSamples(sr)
                for (sample in ls) {
                    val nDotWi = sample.wi!!.dot(sr.normal)
                    if (nDotWi > 0) {
                        var inShadow = false
                        if (light1.shadows) {
                            val shadowRay = Ray(sr.hitPoint, sample.wi!!)
                            inShadow = light1.inShadow(world, shadowRay, sr, sample)
                        }
                        if (!inShadow) {
                            val fd = diffuseBrdf.f(sr, wo, sample.wi!!)
                            val fs = specularBrdf.f(sr, wo, sample.wi!!)
                            val l = light1.L(world, sr, sample)
                            val fsfslndotwi = fd.plus(fs).times(l).times(nDotWi)
                            // TODO: hier ist der Unterschied zu shade()
                            val f1 = light1.G(sr, sample) / light1.pdf(sr)
                            val T = fsfslndotwi.times(f1)
                            S.plus(T)
                        }
                    }
                }
            }
        }
        L = L.plus(S.average)
        return L
    }

    override fun getLe(sr: Shade): Color {
        // TODO
        return specularBrdf.cs!!.getColor(sr).times(specularBrdf.ks)
    }
}

